src/python_testing/spec_parsing_support.py

```diff
@@ -0,0 +1,321 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import os
+import xml.etree.ElementTree as ElementTree
+from copy import deepcopy
+from dataclasses import dataclass
+from enum import Enum, auto
+from typing import Callable
+
+from chip.tlv import uint
+from conformance_support import (DEPRECATE_CONFORM, DISALLOW_CONFORM, MANDATORY_CONFORM, OPTIONAL_CONFORM, OTHERWISE_CONFORM,
+                                 PROVISIONAL_CONFORM, ConformanceDecision, ConformanceParseParameters, or_operation,
+                                 parse_callable_from_xml)
+from matter_testing_support import (AttributePathLocation, ClusterPathLocation, CommandPathLocation, EventPathLocation,
+                                    FeaturePathLocation, ProblemNotice, ProblemSeverity)
+
+
+@dataclass
+class XmlFeature:
+    code: str
+    name: str
+    conformance: Callable[[uint], ConformanceDecision]
+
+
+@dataclass
+class XmlAttribute:
+    name: str
+    datatype: str
+    conformance: Callable[[uint], ConformanceDecision]
+
+
+@dataclass
+class XmlCommand:
+    name: str
+    conformance: Callable[[uint], ConformanceDecision]
+
+
+@dataclass
+class XmlEvent:
+    name: str
+    conformance: Callable[[uint], ConformanceDecision]
+
+
+@dataclass
+class XmlCluster:
+    name: str
+    revision: int
+    derived: str
+    feature_map: dict[str, uint]
+    attribute_map: dict[str, uint]
+    command_map: dict[str, uint]
+    features: dict[str, XmlFeature]
+    attributes: dict[uint, XmlAttribute]
+    accepted_commands: dict[uint, XmlCommand]
+    generated_commands: dict[uint, XmlCommand]
+    events: dict[uint, XmlEvent]
+
+
+class CommandType(Enum):
+    ACCEPTED = auto()
+    GENERATED = auto()
+
+
+def has_zigbee_conformance(conformance: ElementTree.Element) -> bool:
+    # For clusters, things with zigbee conformance can share IDs with the matter elements, so we don't want them
+
+    # TODO: it's actually possible for a thing to have a zigbee conformance AND to have other conformances, and we should check
+    # for that, but for now, this is fine because that hasn't happened in the cluster conformances YET.
+    # It does happen for device types, so we need to be careful there.
+    condition = conformance.iter('condition')
+    for c in condition:
+        try:
+            c.attrib['name'].lower() == "zigbee"
+            return True
+        except KeyError:
+            continue
+    return False
+
+
+class ClusterParser:
+    def __init__(self, cluster, cluster_id, name):
+        self.problems: list[ProblemNotice] = []
+        self.cluster = cluster
+        self.cluster_id = cluster_id
```

Comment:
nit, I think all these fields look to be used internally by `ClusterParser` only, if so does it make sense to use weak “internal use” indicator (ie `_single_leading_underscore`)

Reply:
sure, changed.